fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,431 @@
+                is_lmd_confirmed(store, confirmation_byzantine_threshold, block.parent_root)
+            )
+```
+
+#### `get_remaining_weight_in_epoch`
+
+```python
+def get_remaining_weight_in_epoch(store: Store, current_slot: Slot) -> Gwei:
+    # Returns the total weight of votes for this epoch from future committees after the current slot
+    first_slot_next_epoch = compute_start_slot_at_epoch(Epoch(compute_epoch_at_slot(current_slot) + 1))
+    return get_committee_weight_between_slots(store, Slot(current_slot + 1), Slot(first_slot_next_epoch - 1))
+```
+
+#### `get_leaf_block_roots`
+
+```python
+def get_leaf_block_roots(store: Store, block_root: Root) -> Set[Root]:
+    children = [
+        root for root in store.blocks.keys()
+        if store.blocks[root].parent_root == block_root
+    ]
+
+    if any(children):
+        # Get leaves of all children and add to the set.
+        leaves = set().union(*[get_leaf_block_roots(store, child) for child in children])
+        return leaves
+    else:
+        # This block is a leaf.
+        return set(block_root)
+
+```
+
+#### `get_ffg_support`
+
+```python
+def get_ffg_support(store: Store, block_root: Root) -> Gwei:
+    # Returns the total weight supporting the highest checkpoint in the block's chain
+
+    block = store.blocks[block_root]
+    assert get_current_store_epoch(store) == compute_epoch_at_slot(block.slot)
+
+    leave_roots = get_leaf_block_roots(store, block_root)
+    # current_epoch_attestations contains only attestations with source matching block.current_justified_checkpoint
+    attestations_in_leaves = set().union(
+        *[store.block_states[root].current_epoch_attestations for root in leave_roots]
+    )
+
+    current_epoch = get_current_store_epoch(store)
+    checkpoint_root = get_checkpoint_block(store, block_root, current_epoch)
+    support_for_checkpoint = {a for a in attestations_in_leaves if a.data.target.root == checkpoint_root}
+    checkpoint_state = store.block_states[checkpoint_root]
+    return get_attesting_balance(checkpoint_state, list(support_for_checkpoint))
+```
+
+#### `is_ffg_confirmed`
+
+```python
+def is_ffg_confirmed(
+    # Returns whether the branch will justify it's current epoch checkpoint at the end of this epoch
+    store: Store,
+    confirmation_byzantine_threshold: int,
+    confirmation_slashing_threshold: int,
+    block_root: Root,
+) -> bool:
+    current_slot = get_current_slot(store)
+    block = store.blocks[block_root]
+    assert get_current_store_epoch(store) == compute_epoch_at_slot(block.slot)
+
+    current_epoch = get_current_store_epoch(store)
+    checkpoint_root = get_checkpoint_block(store, block_root, current_epoch)
+    checkpoint_state = store.block_states[checkpoint_root]
+
+    remaining_ffg_weight = int(get_remaining_weight_in_epoch(store, current_slot))
+    total_active_balance = int(get_total_active_balance(checkpoint_state))
```

Comment:
As `checkpoint_root` could refer to a block from a previous epoch, in case of empty slots at end of its epoch (or even entire missed epochs), its checkpoints have been pulled to the current epoch. Is it alright to use the original `total_active_balance`, or do we have to pull that to current epoch as well? And, if we have to pull, why not just use `store.block_states[block_root]` as the basis of `get_total_active_balance`?

Reply:
I think that you are right, we should use the pulled up state.

Using `store.block_states[block_root]` would give us a less live confirmation rule as `remaining_ffg_weight` may get reduced during the course of an epoch due to slashing, which I believe would immediately change a validator's effective balance by means of  `process_attester_slashing` or `process_proposer_slashing` called inside `process_operations`.

Reply:
To make sense of what I have written above, please see [comment below](https://github.com/ethereum/consensus-specs/pull/3339/files#r1230361554)

Reply:
No, Effective balance updates during epoch processing, in `process_effective_balance_updates`.

Slashings also dont affect `get_total_active_balance` immediately, the shufflings and so on remain the same regardless of slashing activity.

Reply:
Thanks for pointing this out.
The code has now been changed to use the block's state to fetch balances.